### DISCUSSION TITLE
Adding support to specify import location via URLs

### DISCUSSION THREAD

**Discussion by didiercrunch** [typst/typst#2658] [Ideas]
Before introducing my formal proposal, allow me to contextualize it.  I have been teaching a couple of classes lately and I have been using typst to write all my documents.  I have many documents in different git repositories[^1].  Usually, multiple typst documents by git repository.  A typical course requires me to write a syllabus, some homeworks and  addition documents; all in typst.   To be professional, these documents need to look the same.  Currently, the only way I have to share the style between documents located in different repositories is to copy typst code.

I know I could leverage git submodules.  As I am using nix as build system, I could use that too.  I am aware (but not up to date) with typst packages.  Nonetheless, I feel my needs would not be fulfilled by any of these technologies.  I would like to have a way to trivially share code between projects without polluting a global repository.

My proposal is to allow imports via URLs in typst in a very similar fashion than [JavaScript modules][jsmodules].  Go uses something similar to import URLs too.  Deno uses JavaScript modules too[^2] and I love my experience with it.  
Below is an example of what I am proposing.

```
#import "https://example.com/didier-typst/common.typ": basic-footer, basic-header
```

Using URLs in imports  would benefits typst  users by allowing them to easily share code between repositories in a very stable manner without polluting a global repository.  For instance, teachers of my department could all use the same template by simply publishing the template on github and referring the template via the url github provides.  For simple documents, a centralize repository requires too much ceremony.  

Some people do not like import URLs because they say it is insecure[^4] and  for the lack of dependency versioning[^5].   Practically, I believe it is possible to fix these issues using vendoring and import maps in a very similar fashion than what deno does [^6]. 

My personal experience of writing a lot of small to medium documents in typst has shown me a need of a clean way to share personal code between projects and I believe adding the possibility import files by path would be great solution.

thanks,

didier

[jsmodules]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Modules

[^1]: https://gitlab.com/didier-ahuntsic
[^2]: https://go.dev/doc/modules/managing-dependencies
[^3]: https://docs.deno.com/runtime/tutorials/manage_dependencies
[^4]: https://deon.land/
[^5]: https://groups.google.com/g/golang-dev/c/nMWoEAG55v8/m/iJGgur7W_SEJ
[^6]: https://docs.deno.com/runtime/manual/basics/import_maps

**Comment by didiercrunch** [comment]
@laurmaedje, @reknih 

Hi there,

I have open this discussion knowing very well that what I am proposing is drastic change in the language.  Nonetheless, I believe this feature could help a lot of typst users in my situation.  From the 5 up vote on this discussion, I assume 4 other person than me would like to see this feature too.

I have a working prof of concept[^1].  It seams fully functional but still lacks couple of things before I can officially open a PR (general clean-up, comments, integration tests, better http integration, ...).

So my question:  would you be willing to merge this kind of feature in typst?

regards,

didier

[^1]: https://github.com/didiercrunch/typst/pull/1

**Reply by didiercrunch** [reply]
Another cool thing about this feature is that we can use jsdelivr as global repository.  For instance, the below code works on my branch.

```typst
import "https://cdn.jsdelivr.net/gh/didiercrunch/typst-repo@main/test-remote/foo.typ": name

This #name comes from a CDN.
```

**Comment by Complex2-Liu** [comment]
I like your proposal! But what if your `common.typ` relies on other typ files, in that case pulling only one typ file is insufficient. Your requirement is similar to mine  #3099 on local package, so that we can put the shared typ files in a git repository, rather than copying the same code for every project.

```typst
#import "@local/common.typ": basic-footer, basic-header
#import "@local/foo.typ": name
```

After taking a look at your `common.typ`, it relies on `main.toml`. I suppose you want to implement something similar to `\usepackage{gitinfo2}` in LaTeX (c.f. [tex stackexchange 450774](https://tex.stackexchange.com/questions/450774))? If so, PR #2894 has already been merged. I am waiting for the next release.

**Reply by laurmaedje** [reply]
I agree that, if at all, remote import only makes sense at the package level. We are considering to add support for importing git repositores in the future. For more discussion, see the [forge thread on Discord](https://discord.com/channels/1054443721975922748/1176122103355953162).

**Reply by didiercrunch** [reply]
@Complex2-Liu 

My branch will fetch assets recursively.  In the below example, I demonstrate how I support transitive dependencies for typ, images and wasm file types.  You can see the result [here](https://github.com/typst/typst/files/13811950/importurl.pdf).  My idea here was to copy deno's behaviour.

```typst
#import "https://cdn.jsdelivr.net/gh/didiercrunch/typst-repo@9795dd7568384064/test-remote/foo.typ": name, type-writer-image
#import "https://raw.githubusercontent.com/typst/packages/main/packages/preview/tiaoma/0.1.0/lib.typ": ean


== Basic example
Below is a value that has been imported:
#name

== Image example
Below is an image that has been imported:
#type-writer-image


== Using package with wasm
Below is the usage of a package via https link.  The package uses wasm.

#ean("1234567890128")
```


@laurmaedje As for the git integration, this would be marvelous.  In many context, you do not want to make your document public before it is ready.  For this, good support of private git repositories is necessary.  

I do not have a good answer for that.  The approach I took is heavily inspired by deno which uses auth tockens to solve this issue[^1].  Combine with import-maps[^2] I am pretty sure users can find good solutions.

An other approach that I am aware of is Go.  I have not looked into the details, but I assume Go sub-processes to git to fetch assets.  It seems this is what you are aiming at. 

Personally, I prefer deno's way over Go's way even if setting environment variables might be too difficult for the average users.

Anyway, I will jump on the discord discussion.

thanks,

didier


[^1]: https://docs.deno.com/runtime/manual/basics/modules/private
[^2]: https://docs.deno.com/runtime/manual/basics/import_maps

**Comment by AwesomeQuest** [comment]
Any updates on when this might be added?
